src/types/mongo/cpt2.rs

```diff
@@ -0,0 +1,436 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::str::FromStr;
+
+use anyhow::{
+    anyhow,
+    bail,
+};
+use mongodb::bson::{
+    doc,
+    Bson,
+    DateTime,
+    Document,
+};
+
+use super::{
+    BsonExt,
+    DocExt,
+};
+use crate::cpt2::{
+    prelude::*,
+    Message,
+};
+
+pub fn message_to_bson(message: &Message) -> Bson {
+    let mut doc = Document::new();
+    doc.insert("protocol_version", 0);
+    doc.insert("network_id", message.network_id() as i64);
+    doc.insert(
+        "parents",
+        message.parents().iter().map(|p| p.to_string()).collect::<Vec<_>>(),
+    );
+    doc.insert("payload", message.payload().as_ref().map(payload_to_bson));
+    doc.insert("nonce", message.nonce().to_string());
+    Bson::Document(doc)
+}
+
+#[allow(unused)]
+pub fn message_from_bson(bson: Bson) -> anyhow::Result<Message> {
+    let doc = bson.to_document()?;
+    message_from_doc(doc)
+}
+
+pub fn message_from_doc(mut doc: Document) -> anyhow::Result<Message> {
+    let mut builder = MessageBuilder::default()
+        .with_network_id(doc.get_i64("network_id")? as u64)
+        .with_parents(Parents::new(
+            doc.take_array("parents")?
+                .into_iter()
+                .map(message_id_from_bson)
+                .collect::<Result<Vec<_>, _>>()?,
+        )?)
+        .with_nonce_provider(doc.get_str("nonce")?.parse::<u64>()?, 0.0);
+    if let Some(payload) = doc.take("payload").ok().map(payload_from_bson).transpose()? {
+        builder = builder.with_payload(payload);
+    }
+    Ok(builder.finish()?)
+}
+
+fn payload_to_bson(payload: &Payload) -> Bson {
+    match payload {
+        Payload::Transaction(t) => transaction_payload_to_bson(&**t),
+        Payload::Milestone(m) => milestone_payload_to_bson(&**m),
+        Payload::Indexation(i) => indexation_payload_to_bson(&**i),
+        Payload::Receipt(r) => receipt_payload_to_bson(&**r),
+        Payload::TreasuryTransaction(t) => treasury_payload_to_bson(&**t),
+    }
+}
+
+pub fn payload_from_bson(bson: Bson) -> anyhow::Result<Payload> {
+    let kind = bson
+        .as_document()
+        .ok_or_else(|| anyhow!("Invalid payload"))?
+        .get_i32("kind")? as u32;
+    Ok(match kind {
+        TransactionPayload::KIND => Payload::Transaction(Box::new(transaction_payload_from_bson(bson)?)),
+        MilestonePayload::KIND => Payload::Milestone(Box::new(milestone_payload_from_bson(bson)?)),
+        IndexationPayload::KIND => Payload::Indexation(Box::new(indexation_payload_from_bson(bson)?)),
+        ReceiptPayload::KIND => Payload::Receipt(Box::new(receipt_payload_from_bson(bson)?)),
+        TreasuryTransactionPayload::KIND => Payload::TreasuryTransaction(Box::new(treasury_payload_from_bson(bson)?)),
+        _ => bail!("Unknown payload kind: {}", kind),
+    })
+}
+
+fn transaction_payload_to_bson(payload: &TransactionPayload) -> Bson {
+    let mut doc = Document::new();
+    doc.insert("kind", TransactionPayload::KIND as i32);
```

Comment:
use `.into()` if possible.

Reply:
Coercing u32 to i32 is fallible operation, so that would look like this:

```
doc.insert("kind", TryInto::<i32>::try_into(TransactionPayload::KIND).unwrap());
```

🤷